### DISCUSSION TITLE
How should I move from sessions to pools?

### DISCUSSION THREAD

**Discussion by javierguzman** [sqlalchemy/sqlalchemy#9918] [Usage Questions] [answered]
Hello all,

I am creating an app using FastAPI with SQLAlchemy and asyncio. However, now I have started getting the following error:
```
cannot perform operation: another operation is in progress
```
Which I have read it is related to the usage of the same session for doing multiple operations.

This is an example of how I operate right now:

```
engine = create_async_engine(
    f"postgresql+asyncpg://{username}:{password}@{host}:{port}/{db_name}"
)

Session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)

async def get_session():
    session = Session()
    async with Session() as session:
        yield session
```
```
    async def add_keywords_to_campaign(
        self, campaign_id: UUID, keywords: List[str]
    ) -> List[CampaignKeyword]:
        try:
            coroutine = [
                self.add_keyword_to_campaign(campaign_id, keywords[k])
                for k in range(len(keywords))
            ]
            keywords_in_campaign = await asyncio.gather(*coroutine)
            await self.session.commit()
            return keywords_in_campaign
        except Exception as err:
            print(err)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Error creating campaign keywords",
            )

    async def add_keyword_to_campaign(
        self, campaign_id: UUID, keyword: str
    ) -> CampaignKeyword:
        try:
            keyword_repository = KeywordRepository(self.session)
            new_keyword = await keyword_repository.create_keyword(keyword)
            new_campaign_keyword = {
                "campaign_id": campaign_id,
                "keyword_id": new_keyword.id,
            }
            statement = (
                insert(CampaignKeyword)
                .values(new_campaign_keyword)
                .on_conflict_do_nothing()
                .returning(CampaignKeyword.campaign_id, CampaignKeyword.keyword_id)
            )
            result = await self.session.execute(statement)
            inserted_campaign_keyword_row = result.one()
            inserted_campaign_keyword = inserted_campaign_keyword_row._asdict()
            return inserted_campaign_keyword
        except exc.SQLAlchemyError as err:
            print(err)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Error adding keyword to campaign",
            )
```         

The above is my campaign repository class which gets initialized with a session using FastAPI depends e.g:
```
@router.patch(
    "/api/campaign/{campaign_id}", tags=["campaign"], status_code=status.HTTP_200_OK
)
async def handle_patch_campaign(
    campaign_id: UUID,
    new_campaign_data: PatchCampaignSchema,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Session = Depends(get_session),
) -> Response[CampaignResponse]:
    campaign_repository = CampaignRepository(session)
    campaign = await campaign_repository.get_campaign_by_id(campaign_id)
```

Could someone point me out into the right direction about how to move from sessions to pools so I can avoid the aforementioned error? I am new into SQLAlchemy and FastAPI so I would appreciate as much hints as possible.

Thank you in advance and regards

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you mean by using the pool. A pool it's not used directly in sqlalchemy.

As the error mentions, you cannot use a session concurrently, so the easiest solution here would be to remove the asyncio gather.
Alternatively each coroutine in the gather should use a different session.

Since you are using insert directly, it may alos be possible to transform that in a single statement, removing the need to run the gather all together

**Reply by javierguzman** [reply]
Thank you for the quick response @CaselIT !  To be honest I just mentioned pools because I saw that error could be fixed by using them, but you mentioned other ways to do that, just a couple of questions:

1. Could you point me out to an example of using insert with multiple values at once? I think this would be the cleanest solution for me and would allow me to remove the gather stuff.

2. If at the end I do not use the insert, remove the asyncio gather and use a different session for each insert, how could I control so that if one of the insert fails everything fail?

Again thank you for your help Federico.

**Reply by CaselIT** [reply, marked as answer]
Regarding 1, you can look at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements

Regarding 2, if you remove the gather, you don't need to use different sessions, just do a loop and run the insert in the loop.
If you want to keep the gather, then you will need to use multiple sessions. In that case each session is independent and has its own db transaction, so if you require a transactional behavior that would not work (at least not without further synchronization, such as using a two phase commit

Overall I would try do use the first option, with a single insert

**Reply by javierguzman** [reply]
Thank you very much again @CaselIT  I am going to give it a try to the bulk insert. However I am curious just for future, if I just remove the gather and make a loop and a commit after the loop. Will I not end up committing before finishing adding stuff? I just ask because I come from Nodejs and it is tricky for me to match my async knowledge from Nodejs to Python.

**Reply by CaselIT** [reply]
I implied that you would await each insert in the for loop, basically making one inser after the other sequentially

**Reply by javierguzman** [reply]
True; then I do not remember why I started using asyncio gather stuff...I might hit the same cause in future if I start removing gather and then I will come back asking about it 🥲

**Reply by CaselIT** [reply]
well using `gather` allows you to run coroutines in parallel, but since a session is ultimately an abstraction over a single connection, you cannot use the same one on parallel coroutines.
the for loop with await will serialize things, so it's fine to use the same session:
```
for stmt in get_insert_statements():
  await session.execute(stmt)
```

**Reply by javierguzman** [reply]
Sorry for coming back this late @CaselIT but I could not test this until now for some personal stuff which made me stop my project. I have resumed now what I was trying to do and I do not manage to use multiple inserts at once. The idea is to assign keywords to a campaign. So first I need to create the keywords and then assign them to a campaign, so far I have replaced the original code with this:
```
def prepare_keywords(self, keywords: List[str]) -> List[Dict]:
        prepared_keywords = []
        for keyword in keywords:
            prepared_keywords.append({"keyword": keyword})
        return prepared_keywords

    async def add_keywords_to_campaign(
        self, campaign_id: UUID, keywords: List[str]
    ) -> List[CampaignKeyword]:
        try:
            print("Keywords to add to campaign")
            prepared_keywords = self.prepare_keywords(keywords)
            print(prepared_keywords)
            keyword_repository = KeywordRepository(self.session)
            statement = (
                insert(Keyword)
                .values(prepared_keywords)
                .on_conflict_do_nothing()
                .returning(Keyword)
            )
            inserted_keywords = await self.session.execute(statement)
            scalar_keywords = inserted_keywords.scalars()
            print(scalar_keywords)
            await self.session.commit()
```
But scalar_keywords returns [] I believe it is because those keywords exist already. It is there a way to insert a bunch of items and return them independently they exist already or not?

Thank you again and regards

**Reply by CaselIT** [reply]
I think returning used with `on_conflict_do_nothing` will only return the inserted rows. This ultimately depends on the db you are using.
If you want back all keywords in the list you are trying to insert then run a select with an condition. But since you already have the keyword on python maybe you don't need do?

**Reply by javierguzman** [reply]
I am using postgresql.   What I pass as an argument is a list of keywords, but they are just strings. I need the ID so I can insert them into the CampaignKeyword table, which is just a table with two columns, campaign ID and keyword ID. I guess that I can run a select after the insert.

**Reply by CaselIT** [reply]
Yes, iirc in postgresql returning in an insert on conflict do nothing includes only the inserted rows. You would have to check the postgresql docs for more details

**Reply by javierguzman** [reply]
Thank you very much for your help @CaselIT I think I have managed to solve it, I leave the code with comments added in case it helps someone in future:

```py
      async def get_keywords_from_list(self, keywords: List[str]) -> List[Keyword]:
        try:
            query = select(Keyword).filter(Keyword.keyword.in_(keywords))
            result = await self.session.execute(query)
            fetched_keywords = result.scalars().all()
            return fetched_keywords
        except exc.SQLAlchemyError as err:
            print(err)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Error getting keywords from list",
            )
            
    def prepare_campaign_keywords(
        self, campaign_id: str, keywords: List[Dict]
    ) -> List[Dict]:
        prepared_campaign_keywords = []
        for keyword in keywords:
            prepared_campaign_keywords.append(
                {"campaign_id": campaign_id, "keyword_id": keyword.id}
            )
        return prepared_campaign_keywords

    async def add_keywords_to_campaign(
        self, campaign_id: UUID, keywords: List[str]
    ) -> List[Keyword]:
        try:
            keyword_repository = KeywordRepository(self.session)
            await keyword_repository.create_keywords(keywords) #we add keywords to keywords table
            campaign_keywords = await keyword_repository.get_keywords_from_list(
                keywords
            )  # we query the keywords we just inserted so we can get the ID
            prepared_campaign_keywords = self.prepare_campaign_keywords(
                campaign_id, campaign_keywords
            )
            statement = (
                insert(CampaignKeyword)
                .values(prepared_campaign_keywords)
                .on_conflict_do_nothing()
            )
            await self.session.execute(statement)
            await self.session.commit()
            campaign_keywords = await self.get_keywords_from_campaign(campaign_id) #we fetch all keywords after insert
            return campaign_keywords
        except Exception as err:
            print(err)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Error adding campaign keywords",
            )
```
